### DISCUSSION TITLE
is there a way to process n number of files at a time?

### DISCUSSION THREAD

**Discussion by complexgenome** [nextflow-io/nextflow#4083] [Q&A] [answered]
Hello there,

I've 11 samples (11 tumors, 11 normals) that I'd like to process using nextflow pipeline I've put. At the moment all 11 are processed in parallel. Is there a way that can restrict the number of samples to process in parallel? For example process 4 at a time, then next 4 then remaining, in this case 3.

I have code as:
` read_pairs_ch = Channel.fromFilePairs( params.raw_files).take(3) `

But it only takes 3 sample, remaining are not processed. I'd like to have three-three at a time.
How do I enable processing three samples at a time?

**Comment by bentsherman** [comment]
You can use the [collate](https://nextflow.io/docs/edge/operator.html#collate) operator to group the samples into groups of 4 (including the remainder group):
```groovy
read_pairs_ch = Channel.fromFilePairs( params.raw_files).collate(4)
```

Alternatively, you can use the [maxForks](https://nextflow.io/docs/edge/process.html#maxforks) directive to limit the number of parallel tasks.

**Comment by complexgenome** [comment]
Thanks for your reply. When I put collate, I run into error for `WARN: Input tuple does not match input set cardinality declared by process`

Please see attached process, main.nf scripts along with error.
<img width="1642" alt="Screen Shot 2023-07-07 at 12 58 06 PM" src="https://github.com/nextflow-io/nextflow/assets/5683969/5d01a8df-d2ca-4ac4-a71e-7e81517c4700">
<img width="993" alt="Screen Shot 2023-07-07 at 12 58 24 PM" src="https://github.com/nextflow-io/nextflow/assets/5683969/f1df9416-8d37-4f1b-902f-7aceaca694f4">
<img width="1007" alt="Screen Shot 2023-07-07 at 12 58 50 PM" src="https://github.com/nextflow-io/nextflow/assets/5683969/4ac796cd-11fe-404e-a889-9c1a9be67016">

**Reply by bentsherman** [reply, marked as answer]
Yes, this is because your process expects a single sample, whereas collate groups multiple samples into a tuple. This is why I suggested `maxForks` as the better way to limit the parallelism.

**Comment by complexgenome** [comment]
Understood, thank you 🙏 so much.
